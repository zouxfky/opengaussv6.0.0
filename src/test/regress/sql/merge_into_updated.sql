create schema merge_into_updated;
set search_path = 'merge_into_updated';

create table t1 (c1 int, c2 text, c3 timestamp);
insert into t1 values (1, 'a', '2023-09-15');

create or replace function t1_tri_func() return trigger as
begin
    new.c3 = '2023-09-16';
    return new;
end;
/
create trigger t1_tri
    before update on t1
    for each row
    execute procedure t1_tri_func();

-- success, matched, do update
\parallel on 2
begin
    update t1 set c2 = 'b' where c1 = 1;
    perform pg_sleep(3);
end;
/
begin
    perform pg_sleep(1);
    merge into t1 using (select 1 c1) t2
    on (t1.c1 = t2.c1)
    when matched then update set c2 = 'c'
    when not matched then insert values (2, 'c', '2023-09-17');
end;
/
\parallel off
select c1, c2, to_char(c3, 'yyyy-mm-dd') from t1 order by c1;

-- success, matched, do update
\parallel on 2
begin
    update t1 set c2 = 'b' where c1 = 1;
    perform pg_sleep(3);
end;
/
begin
    perform pg_sleep(1);
    insert into t1 values (1, 'hello', '2023-09-17');
    delete from t1 where c2 = 'hello';
    merge into t1 using (select 1 c1) t2
    on (t1.c1 = t2.c1)
    when matched then update set c2 = 'd'
    when not matched then insert values (2, 'c', '2023-09-17');
end;
/
\parallel off
select c1, c2, to_char(c3, 'yyyy-mm-dd') from t1 order by c1;

-- success, concurrently update join condition, not matched, do insert
\parallel on 2
begin
    update t1 set c1 = 2 where c1 = 1;
    perform pg_sleep(3);
end;
/
begin
    perform pg_sleep(1);
    merge into t1 using (select 1 c1) t2
    on (t1.c1 = t2.c1)
    when matched then update set c2 = 'd'
    when not matched then insert values (3, 'd', '2023-09-18');
end;
/
\parallel off
select c1, c2, to_char(c3, 'yyyy-mm-dd') from t1 order by c1;

-- success, not matched, do insert
\parallel on 2
begin
    update t1 set c2 = 'b';
    perform pg_sleep(3);
end;
/
begin
    perform pg_sleep(1);
    insert into t1 values (1, 'hello', '2023-09-17');
    delete from t1;
    merge into t1 using (select 1 c1) t2
    on (t1.c1 = t2.c1)
    when matched then update set c2 = 'c'
    when not matched then insert values (2, 'c', '2023-09-17');
end;
/
\parallel off
select c1, c2, to_char(c3, 'yyyy-mm-dd') from t1 order by c1;

-- trigger update the join condition, not matched, do insert
create or replace function t1_tri_func() return trigger as
begin
    new.c1 = 4;
    return new;
end;
/
\parallel on 2
begin
    update t1 set c2 = 'b';
    perform pg_sleep(3);
end;
/
begin
    perform pg_sleep(1);
    merge into t1 using (select 2 c1) t2
    on (t1.c1 = t2.c1)
    when matched then update set c2 = 'e'
    when not matched then insert values (4, 'e', '2023-09-19');
end;
/
\parallel off
select c1, c2, to_char(c3, 'yyyy-mm-dd') from t1 order by c1;

-- materialized view
drop table if exists t_a;
drop table if exists t_b;
create table t_a(
a_id int,
a_name varchar2(100)
);
create table t_b(
b_id int,
b_name varchar2(100)
);
DROP materialized VIEW if exists v_t_b;
create materialized view v_t_b as select * from t_b;

MERGE INTO t_a a  
USING v_t_b vb
on (a.a_id=vb.b_id)
WHEN MATCHED THEN
update set a.a_name=vb.b_name;

DROP materialized VIEW v_t_b;
drop table t_a;
drop table t_b;

drop schema merge_into_updated cascade;
