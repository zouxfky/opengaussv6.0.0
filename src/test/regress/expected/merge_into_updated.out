create schema merge_into_updated;
set search_path = 'merge_into_updated';
create table t1 (c1 int, c2 text, c3 timestamp);
insert into t1 values (1, 'a', '2023-09-15');
create or replace function t1_tri_func() return trigger as
begin
    new.c3 = '2023-09-16';
    return new;
end;
/
create trigger t1_tri
    before update on t1
    for each row
    execute procedure t1_tri_func();
-- success, matched, do update
\parallel on 2
begin
    update t1 set c2 = 'b' where c1 = 1;
    perform pg_sleep(3);
end;
/
begin
    perform pg_sleep(1);
    merge into t1 using (select 1 c1) t2
    on (t1.c1 = t2.c1)
    when matched then update set c2 = 'c'
    when not matched then insert values (2, 'c', '2023-09-17');
end;
/
\parallel off
select c1, c2, to_char(c3, 'yyyy-mm-dd') from t1 order by c1;
 c1 | c2 |  to_char   
----+----+------------
  1 | c  | 2023-09-16
(1 row)

-- success, matched, do update
\parallel on 2
begin
    update t1 set c2 = 'b' where c1 = 1;
    perform pg_sleep(3);
end;
/
begin
    perform pg_sleep(1);
    insert into t1 values (1, 'hello', '2023-09-17');
    delete from t1 where c2 = 'hello';
    merge into t1 using (select 1 c1) t2
    on (t1.c1 = t2.c1)
    when matched then update set c2 = 'd'
    when not matched then insert values (2, 'c', '2023-09-17');
end;
/
\parallel off
select c1, c2, to_char(c3, 'yyyy-mm-dd') from t1 order by c1;
 c1 | c2 |  to_char   
----+----+------------
  1 | d  | 2023-09-16
(1 row)

-- success, concurrently update join condition, not matched, do insert
\parallel on 2
begin
    update t1 set c1 = 2 where c1 = 1;
    perform pg_sleep(3);
end;
/
begin
    perform pg_sleep(1);
    merge into t1 using (select 1 c1) t2
    on (t1.c1 = t2.c1)
    when matched then update set c2 = 'd'
    when not matched then insert values (3, 'd', '2023-09-18');
end;
/
\parallel off
select c1, c2, to_char(c3, 'yyyy-mm-dd') from t1 order by c1;
 c1 | c2 |  to_char   
----+----+------------
  2 | d  | 2023-09-16
  3 | d  | 2023-09-18
(2 rows)

-- success, not matched, do insert
\parallel on 2
begin
    update t1 set c2 = 'b';
    perform pg_sleep(3);
end;
/
begin
    perform pg_sleep(1);
    insert into t1 values (1, 'hello', '2023-09-17');
    delete from t1;
    merge into t1 using (select 1 c1) t2
    on (t1.c1 = t2.c1)
    when matched then update set c2 = 'c'
    when not matched then insert values (2, 'c', '2023-09-17');
end;
/
\parallel off
select c1, c2, to_char(c3, 'yyyy-mm-dd') from t1 order by c1;
 c1 | c2 |  to_char   
----+----+------------
  2 | c  | 2023-09-17
(1 row)

-- trigger update the join condition, not matched, do insert
create or replace function t1_tri_func() return trigger as
begin
    new.c1 = 4;
    return new;
end;
/
\parallel on 2
begin
    update t1 set c2 = 'b';
    perform pg_sleep(3);
end;
/
begin
    perform pg_sleep(1);
    merge into t1 using (select 2 c1) t2
    on (t1.c1 = t2.c1)
    when matched then update set c2 = 'e'
    when not matched then insert values (4, 'e', '2023-09-19');
end;
/
\parallel off
select c1, c2, to_char(c3, 'yyyy-mm-dd') from t1 order by c1;
 c1 | c2 |  to_char   
----+----+------------
  4 | b  | 2023-09-17
  4 | e  | 2023-09-19
(2 rows)

-- materialized view
drop table if exists t_a;
NOTICE:  table "t_a" does not exist, skipping
drop table if exists t_b;
NOTICE:  table "t_b" does not exist, skipping
create table t_a(
a_id int,
a_name varchar2(100)
);
create table t_b(
b_id int,
b_name varchar2(100)
);
DROP materialized VIEW if exists v_t_b;
NOTICE:  materialized view "v_t_b" does not exist, skipping
create materialized view v_t_b as select * from t_b;
MERGE INTO t_a a  
USING v_t_b vb
on (a.a_id=vb.b_id)
WHEN MATCHED THEN
update set a.a_name=vb.b_name;
DROP materialized VIEW v_t_b;
drop table t_a;
drop table t_b;
drop schema merge_into_updated cascade;
NOTICE:  drop cascades to 2 other objects
DETAIL:  drop cascades to table t1
drop cascades to function t1_tri_func()
